### PR TITLE
Implement mesh items

### PR DESCRIPTION
**Pull request by sapier**
just add 
mesh = "character.b3d"
meshtexture = "character.png"

to a craftitem definition and replace it's inventory image by a plain texture.

I called it "mesh" as this commit should show up where to change to add support for using arbitrary 3d models as items.

``` lua
 minetest.register_craftitem("default:steel_ingot", {
    description = "Steel Ingot",
    inventory_image = "default_steel_ingot.png",
    mesh = "character.b3d",
    meshtexture = "character.png",
 })
```

http://imgur.com/y9ixW4r http://imgur.com/xYEUIzs http://imgur.com/z4hlJgb

@all I'm thinking about removing the ingot special code as same can be done by adding a ingot mesh too. It's kindof redundant now.

### CODE REVIEW

**Comment by paramat** [comment]
:-1: I feel this looks inconsistent and out of place amongst all other items which are simple, pixelated and unrealistic extrusions of inventory textures. I feel the appearence of a dropped item is not that important so this seems excessively fancy.

**Comment by sapier** [comment]
Hmm I guess the screenshot ain't clear enough. 
This ain't primary about the dropped item it's about:

*Inventory Item
*Wield Item
*ans dropped Item

This screenshot shows old and new style drawn ingots both with 16x16 textures.
http://imgur.com/xYEUIzs

The dropped item is the part I cared least, more important to me have been wield item and most important inventory item.

Most effective change will occur in 3d-mode where ingots will look the way nodes look like atm, especially for wield item this is a great improvement.

**Comment by Calinou** [comment]
:-1:, those look like placeable nodes. Moreover, it will break consistency with existing mods (which will have to be updated).

**Comment by sapier** [comment]
Calinou there ain't any breakage with existing mods, noone has to use that feature.
Goddamn could you guys even try it before claiming "breaks things"
If you don't change your game you won't even see ANY difference.

**Comment by kilbith** [comment]
> Calinou there ain't any breakage with existing mods, noone has to use that feature.

He did say breaking the **consistency** with existing mods, i.e. some ingots will stay pixelated, some others as meshes - this is inconsistent.

**Comment by sapier** [comment]
@kilbith that's happening on any texture set so where's the point?

**Comment by est31** [comment]
The drawtype is specified by the `drawtype` key not the `mesh` key.

**Comment by sapier** [comment]
"The drawtype is specified by the drawtype key not the mesh key. "  -->fixed broken pull naming, it's not adding a new drawtype but a separate type of item

File: src/content_cao.cpp [hunk 1]
```diff
@@ -954,15 +954,61 @@ void GenericCAO::addToScene(scene::ISceneManager *smgr, ITextureSource *tsrc,
...
+						m_prop.visual_size.X));
+				u8 li = m_last_light;
+				setMeshColor(m_meshnode->getMesh(), video::SColor(255,li,li,li));
+
+				m_meshnode->setMaterialFlag(video::EMF_LIGHTING, false);
+				m_meshnode->setMaterialFlag(video::EMF_BILINEAR_FILTER, false);
+				m_meshnode->setMaterialType(video::EMT_TRANSPARENT_ALPHA_CHANNEL_REF);
+				m_meshnode->setMaterialFlag(video::EMF_FOG_ENABLE, true);
+			}
+			else if (item.getDefinition(idef).meshname != "")
```
Comment (nerzhul) [inline src/content_cao.cpp]: please use !str.empty() which is faster than this call

**Comment by HybridDog** [comment]
Generating inventory cubes takes a lot time, so l always need to wait for the inventory page to load.
Does generating ingot meshs also take so much time?

**Comment by sapier** [comment]
Same mechanism as inventory cubes so same time yes.
But this issue could be fixed by the "fancy items" setting too for ppl having that old machines .... thought my 8 year old pc is old ;)

**Comment by HybridDog** [comment]
my pc is not old and l have to wait
How about rendering fancy inventory images in another thread and show dummies first?

**Comment by sapier** [comment]
HybridDog did you disable preload item visuals? if you did how long does it take to render a single image for you? If it's more then a fraction of a second there's something else wrong.
What pc do you have?

**Comment by HybridDog** [comment]
I think it doesn't take longer than a second, but it also takes more than 0.1 s, which is very slow for calculating simple cube images.

**Comment by sapier** [comment]
Actually it's more then that it's rendering each single cube to a texture rereading it from gpu ... quite ugly solution. We could render multiple cubes to same texture but especially for older gpus texture size is limited too so we would have lot of work and in worst case no real benefit

**Comment by paramat** [comment]
My opinion is that the 'true form' of items and tools *is* a low-res 2D pixel-art texture which is then extruded as wield and as a dropped item, so there is no need to change this to a mesh, as if the current visuals are an imperfect representation that needs to be improved. It's crude character is not a problem and very few complain.

For example the wielded Minecraft tool has become a much-loved symbol of the game even though it is a low-res extruded 2D texture. The more HD and realistic items become the less character they have and the less lovable they are.

Your ingot mesh is actually too realistic even as a node placed in the world, as it has gently sloping sides. It's also a specific single-use addition that for consistency would need to lead to adding complex meshes for every type of item.

But it looks like you've just removed that. I'm not opposed to adding the option of using a mesh for an item in mods or subgames.

**Comment by HybridDog** [comment]
>The more HD and realistic items become the less character they have and the less lovable they are.

You could use nodebox instead of diagonals.

**Comment by sapier** [comment]
I'm not gonna take part on philosophic discussions about "true form" or not this pull just provides additional functions which don't cause ANY sideeffects if you don't use them.

Imho your "true form" graphics just look weired for a lot of things. And you don't seem to have issues with nodes not beeing 3d extruded items.

**Comment by kaeza** [comment]
:+1:

This allows more possibilities. Whether or not to use them is up to game/mod authors.

By the same "inconsistent" logic, why don't we remove meshes, and disallow textures larger than 16px or with more than X unique colors?

More flexibility is good. Don't you think?

**Comment by est31** [comment]
I think this is doable by mods already, isn't it? I don't think we should introduce additional complexity if you can do it with another way already as well.

**Comment by est31** [comment]
And if there is enough adoption in servers etc then we can add it to the engine as well. Note that once added, we have to support the thing for eternity, it affects all levels, modding and protocol.

**Comment by kaeza** [comment]
It is "possible", but frankly, looks like shit most of the time.
As an aside, could you add support for custom wieldmeshes independent of inventorymeshes?

**Comment by paramat** [comment]
>  I'm thinking about removing the ingot special code as same can be done by adding a ingot mesh too.

Yes this seems better, more general. I agree with what c55 wrote in IRC.

> you don't seem to have issues with nodes not being 3d extruded items.

Those are nodes and nodeboxes so are different, as c55 wrote in IRC:
"cubes are a sane specialization because the base world consists of cubes
there's nothing else the base world consists of, however; certainly not of ingots"
Also nodes would be difficult to recognise if reduced to a 16x16 inventory image, items are smaller and simpler and their initial form is a 16x16 inventory image so no detail is lost.

**Comment by HybridDog** [comment]
Would you like 3d ingot inventory images if the ingots rotate when the mouse cursor is over them like in warzone2100?

**Comment by RealBadAngel** [comment]
this is imho kinda wrong approach, instead item shall be given "mesh" field to be used when provided.
Then formspecs shall be changed to show actual ingame meshes - without the need to build inventory textures (see minimap, its displaying actual, flat meshes, so its quite easy).

**Comment by BlockMen** [comment]
New PR for this: #3479

@sapier plz add a link in situaions like this, so that it is clear for everyone to be not canceled completely

**Pull request closed without merging**